<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Eric Archer" />

<meta name="date" content="2020-02-03" />

<title>Creating and Manipulating gtypes Objects</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Creating and Manipulating gtypes Objects</h1>
<h4 class="author">Eric Archer</h4>
<h4 class="date">2020-02-03</h4>



<div id="raw-data" class="section level2">
<h2>Raw data</h2>
<p>It is easiest to load raw genotype data from the disk if it is available as a text file, usually in comma-delimited (.csv) format. The standard R functions <em>read.table</em> or <em>read.csv</em> can be used to accomplish this. However, in <em>strataG</em>, the <em>readGenData</em> function has been provided for .csv files, which is a wrapper for <em>read.csv</em> that sets commonly used values for missing data and removes blank lines.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gen.data &lt;-<span class="st"> </span><span class="kw">readGenData</span>(<span class="st">&quot;msats.csv&quot;</span>)</code></pre></div>
<pre><code>## Warning in data.table::fread(file = file, header = TRUE, na.strings =
## na.strings, : na.strings[4]==&quot; &quot; consists only of whitespace, ignoring.
## strip.white==TRUE (default) and &quot;&quot; is present in na.strings, so any number of
## spaces in string columns will already be read as &lt;NA&gt;.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(gen.data)</code></pre></div>
<pre><code>## 'data.frame':    126 obs. of  12 variables:
##  $ ids    : chr  &quot;4495&quot; &quot;4496&quot; &quot;4498&quot; &quot;5814&quot; ...
##  $ strata : chr  &quot;Offshore.North&quot; &quot;Offshore.North&quot; &quot;Offshore.North&quot; &quot;Offshore.North&quot; ...
##  $ D11t.1 : chr  &quot;133&quot; &quot;137&quot; &quot;133&quot; &quot;135&quot; ...
##  $ D11t.2 : chr  &quot;133&quot; &quot;141&quot; &quot;133&quot; &quot;139&quot; ...
##  $ EV37.1 : chr  NA NA &quot;220&quot; &quot;210&quot; ...
##  $ EV37.2 : chr  NA NA &quot;220&quot; &quot;226&quot; ...
##  $ EV94.1 : chr  NA &quot;243&quot; &quot;245&quot; &quot;251&quot; ...
##  $ EV94.2 : chr  NA &quot;251&quot; &quot;265&quot; &quot;265&quot; ...
##  $ Ttr11.1: chr  &quot;211&quot; NA &quot;209&quot; &quot;209&quot; ...
##  $ Ttr11.2: chr  &quot;211&quot; NA &quot;211&quot; &quot;209&quot; ...
##  $ Ttr34.1: chr  &quot;185&quot; &quot;189&quot; &quot;189&quot; &quot;185&quot; ...
##  $ Ttr34.2: chr  &quot;191&quot; &quot;189&quot; &quot;189&quot; &quot;187&quot; ...
##  - attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt;</code></pre>
<p>For sequence data stored in FASTA format, the <em>read.fasta</em> function is available, which is a wrapper for the <em>read.dna</em> function in the <em>ape</em> package with standard FASTA arguments set. This will create a DNAbin object in the workspace:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fname &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata/dolph.seqs.fasta&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;strataG&quot;</span>)
x &lt;-<span class="st"> </span><span class="kw">read.fasta</span>(fname) 
x</code></pre></div>
<pre><code>## 126 DNA sequences in binary format stored in a list.
## 
## All sequences of same length: 402 
## 
## Labels:
## 4495
## 4496
## 4498
## 5814
## 5815
## 5816
## ...
## 
## Base composition:
##     a     c     g     t 
## 0.301 0.229 0.129 0.341 
## (Total: 50.65 kb)</code></pre>
<p>For sequences stored in other formats, <em>read.dna</em> should be used directly.</p>
</div>
<div id="construction" class="section level2">
<h2>Construction</h2>
<p>For most functions in <em>strataG</em>, you will need to load your data into a <em>gtypes</em> object. A <em>gtypes</em> object is an R S4 class with several slots that are fully described in <code>?gtypes</code>.<br />
The easiest way to create a <em>gtypes</em> object is with the <em>df2gtypes()</em> function. This function assumes that you have a matrix or data.frame with columns for individual ids, stratification, and locus data. You then specify the columns in the data.frame where this information can be found. <em>df2gtypes()</em> can be used for data with multiple alleles per locus, like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create a single data.frame with the msat data and stratification</span>
msats.merge &lt;-<span class="st"> </span><span class="kw">merge</span>(dolph.strata, dolph.msats, <span class="dt">all.y =</span> <span class="ot">TRUE</span>, <span class="dt">description =</span> <span class="kw">date</span>())
<span class="kw">str</span>(msats.merge)</code></pre></div>
<pre><code>## 'data.frame':    126 obs. of  14 variables:
##  $ id     : chr  &quot;18650&quot; &quot;18651&quot; &quot;18652&quot; &quot;18653&quot; ...
##  $ dLoop  : chr  &quot;Hap.06&quot; &quot;Hap.14&quot; &quot;Hap.23&quot; &quot;Hap.24&quot; ...
##  $ broad  : chr  &quot;Offshore&quot; &quot;Offshore&quot; &quot;Offshore&quot; &quot;Offshore&quot; ...
##  $ fine   : chr  &quot;Offshore.South&quot; &quot;Offshore.South&quot; &quot;Offshore.South&quot; &quot;Offshore.South&quot; ...
##  $ D11t.1 : chr  &quot;131&quot; &quot;137&quot; &quot;131&quot; &quot;133&quot; ...
##  $ D11t.2 : chr  &quot;133&quot; &quot;143&quot; &quot;133&quot; &quot;139&quot; ...
##  $ EV37.1 : chr  &quot;212&quot; &quot;200&quot; &quot;212&quot; &quot;202&quot; ...
##  $ EV37.2 : chr  &quot;222&quot; &quot;228&quot; &quot;218&quot; &quot;220&quot; ...
##  $ EV94.1 : chr  &quot;263&quot; &quot;249&quot; &quot;249&quot; &quot;229&quot; ...
##  $ EV94.2 : chr  &quot;265&quot; &quot;251&quot; &quot;251&quot; &quot;245&quot; ...
##  $ Ttr11.1: chr  &quot;197&quot; &quot;197&quot; &quot;211&quot; &quot;197&quot; ...
##  $ Ttr11.2: chr  &quot;209&quot; &quot;197&quot; &quot;213&quot; &quot;215&quot; ...
##  $ Ttr34.1: chr  &quot;185&quot; &quot;185&quot; &quot;185&quot; &quot;189&quot; ...
##  $ Ttr34.2: chr  &quot;187&quot; &quot;187&quot; &quot;191&quot; &quot;195&quot; ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create the gtypes object</span>
msats.fine &lt;-<span class="st"> </span><span class="kw">df2gtypes</span>(msats.merge, <span class="dt">ploidy =</span> <span class="dv">2</span>, <span class="dt">id.col =</span> <span class="dv">1</span>, <span class="dt">strata.col =</span> <span class="dv">3</span>, <span class="dt">loc.col =</span> <span class="dv">5</span>)</code></pre></div>
<p>…or for haploid data, like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(dolph.seqs)

seq.df &lt;-<span class="st"> </span>dolph.strata[ <span class="kw">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;broad&quot;</span>, <span class="st">&quot;id&quot;</span>)]
<span class="kw">colnames</span>(seq.df)[<span class="dv">3</span>] &lt;-<span class="st"> &quot;D-loop&quot;</span>
dl.g &lt;-<span class="st"> </span><span class="kw">df2gtypes</span>(seq.df, <span class="dt">ploidy =</span> <span class="dv">1</span>, <span class="dt">sequences =</span> dolph.seqs)
dl.g</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 126 samples, 1 locus, 2 strata
## 
## Strata summary:
##    stratum num.ind num.missing num.haplotypes
## 1  Coastal      68           0             68
## 2 Offshore      58           0             58
## 
## Sequence summary:
##    locus num.seqs mean.length mean.num.ns mean.num.indels
## 1 D-loop      126         402           0               0</code></pre>
<p>Note that since each sequence in <em>dolph.seqs</em> is for a given individual, the <em>num.ind</em> and <em>num.haplotypes</em> values are the same for both strata. In order to convert the sequences to unique haplotypes, use the <em>labelHaplotypes()</em> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dl.haps &lt;-<span class="st"> </span><span class="kw">labelHaplotypes</span>(dl.g)
dl.haps</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 126 samples, 1 locus, 2 strata
## Other info: haps.unassigned
## 
## Strata summary:
##    stratum num.ind num.missing num.haplotypes
## 1  Coastal      68           0              5
## 2 Offshore      58           0             29
## 
## Sequence summary:
##    locus num.seqs mean.length mean.num.ns mean.num.indels
## 1 D-loop       33         402           0               0</code></pre>
<div id="sequence2gtypes---convert-dna-sequences" class="section level3">
<h3><em>sequence2gtypes</em> - Convert DNA sequences</h3>
<p>The <em>sequence2gtypes()</em> function creates an unstratified <em>gtype</em> object with just a set of DNA sequences:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(dolph.haps)

haps.g &lt;-<span class="st"> </span><span class="kw">sequence2gtypes</span>(dolph.haps)
haps.g</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 33 samples, 1 locus, 1 stratum
## 
## Strata summary:
##   stratum num.ind num.missing num.haplotypes
## 1 Default      33           0             33
## 
## Sequence summary:
##   locus num.seqs mean.length mean.num.ns mean.num.indels
## 1 gene1       33         402           0               0</code></pre>
<p>If you have a vector that identifies strata designations for the sequences, that can be supplied as well:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract and name the stratification scheme</span>
strata &lt;-<span class="st"> </span>dolph.strata<span class="op">$</span>fine
<span class="kw">names</span>(strata) &lt;-<span class="st"> </span>dolph.strata<span class="op">$</span>ids

<span class="co"># create the gtypes object</span>
dloop.fine &lt;-<span class="st"> </span><span class="kw">sequence2gtypes</span>(dolph.seqs, strata, <span class="dt">seq.names =</span> <span class="st">&quot;dLoop&quot;</span>,
  <span class="dt">description =</span> <span class="st">&quot;dLoop: fine-scale stratification&quot;</span>)
dloop.fine</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dLoop: fine-scale stratification &gt;&gt;&gt;
## 
## Contents: 126 samples, 1 locus, 3 strata
## 
## Strata summary:
##          stratum num.ind num.missing num.haplotypes
## 1        Coastal      68           0             68
## 2 Offshore.North      40           0             40
## 3 Offshore.South      18           0             18
## 
## Sequence summary:
##   locus num.seqs mean.length mean.num.ns mean.num.indels
## 1 dLoop      126         402           0               0</code></pre>
<p>Note that stratification is generally provided for individuals, thus if you want to stratify the resulting <em>gtypes</em> object from <em>sequence2gtypes()</em>, one sequence for each individual should be provided, rather than just a set of unique haplotypes.</p>
</div>
<div id="conversions-from-other-packages." class="section level3">
<h3>Conversions from other packages.</h3>
<p>THere are conversion functions for data objects from several other popular packages in R, such as <em>adegenet</em>(genind), <em>pegas</em>(loci), and <em>phangorn</em>(phydat).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(adegenet)</code></pre></div>
<pre><code>## Loading required package: ade4</code></pre>
<pre><code>## 
##    /// adegenet 2.1.2 is loaded ////////////
## 
##    &gt; overview: '?adegenet'
##    &gt; tutorials/doc/questions: 'adegenetWeb()' 
##    &gt; bug reports/feature requests: adegenetIssues()</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># from example(df2genind)</span>
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">locusA=</span><span class="kw">c</span>(<span class="st">&quot;11&quot;</span>,<span class="st">&quot;11&quot;</span>,<span class="st">&quot;12&quot;</span>,<span class="st">&quot;32&quot;</span>),
                 <span class="dt">locusB=</span><span class="kw">c</span>(<span class="ot">NA</span>,<span class="st">&quot;34&quot;</span>,<span class="st">&quot;55&quot;</span>,<span class="st">&quot;15&quot;</span>),
                 <span class="dt">locusC=</span><span class="kw">c</span>(<span class="st">&quot;22&quot;</span>,<span class="st">&quot;22&quot;</span>,<span class="st">&quot;21&quot;</span>,<span class="st">&quot;22&quot;</span>))
<span class="kw">row.names</span>(df) &lt;-<span class="st"> </span><span class="kw">.genlab</span>(<span class="st">&quot;genotype&quot;</span>,<span class="dv">4</span>)
obj &lt;-<span class="st"> </span><span class="kw">df2genind</span>(df, <span class="dt">ploidy=</span><span class="dv">2</span>, <span class="dt">ncode=</span><span class="dv">1</span>)
obj</code></pre></div>
<pre><code>## /// GENIND OBJECT /////////
## 
##  // 4 individuals; 3 loci; 9 alleles; size: 6.5 Kb
## 
##  // Basic content
##    @tab:  4 x 9 matrix of allele counts
##    @loc.n.all: number of alleles per locus (range: 2-4)
##    @loc.fac: locus factor for the 9 columns of @tab
##    @all.names: list of allele names for each locus
##    @ploidy: ploidy of each individual  (range: 2-2)
##    @type:  codom
##    @call: df2genind(X = df, ncode = 1, ploidy = 2)
## 
##  // Optional content
##    - empty -</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert to gtypes</span>
gi.g &lt;-<span class="st"> </span><span class="kw">genind2gtypes</span>(obj)
gi.g</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:56 &gt;&gt;&gt;
## 
## Contents: 4 samples, 3 loci, 1 stratum
## Other info: genind
## 
## Strata summary:
##   stratum num.ind num.missing num.alleles
## 1 Default       4   0.3333333           3</code></pre>
</div>
</div>
<div id="accessor-functions" class="section level2">
<h2>Accessor functions</h2>
<p>There are several functions for getting basic information from a <em>gtypes</em> object (see <code>?accessors</code>):</p>
<ul>
<li><strong>getNumInd(g)</strong> The number of individuals.</li>
<li><strong>getNumLoci(g)</strong> The number of loci.</li>
<li><strong>getNumStrata(g)</strong> The number of strata in the current scheme.</li>
<li><strong>getIndNames(g)</strong> The names of the individuals.</li>
<li><strong>getLociNames(g)</strong> The names of the loci or genes.</li>
<li><strong>getStrataNames(g)</strong> The names of the strata in the current scheme.</li>
<li><strong>getPloidy(g)</strong> The ploidy of each locus.</li>
<li><strong>getStrata(g)</strong> The current strata to which each individual belongs.</li>
<li><strong>getSchemes(g)</strong> A data frame of potential stratification schemes.</li>
<li><strong>getSequences(g)</strong> The sequences stored in a haploid object.</li>
<li><strong>getDescription(g)</strong> The text description of the object.</li>
<li><strong>getOther(g)</strong> The list used to store other information about the object.</li>
</ul>
<p>Some functions are available for modifying values in the object as well, such as:</p>
<ul>
<li><strong>setStrata(g)</strong> Replace the vector of strata assignments.</li>
<li><strong>setSchemes(g)</strong> Replace the data.frame of potential stratification schemes.</li>
<li><strong>setDescription(g)</strong> Replace the label describing the object.</li>
<li><strong>setOther(g)</strong> Replace the optional data stored in the @other slot.</li>
</ul>
</div>
<div id="subsettingindexing" class="section level2">
<h2>Subsetting/Indexing</h2>
<p>A <em>gtypes</em> object can be subset using the standard R ‘[’ indexing operation, with three slots: [i, j, k]. The first (i) specifies the desired individuals, the second (j) is the loci to return, and the third (k) is the strata. All standard R indexing operations involving numerical, character, or logical vectors work for each argument. For example, to return 10 random individuals:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>msats.fine[<span class="kw">sample</span>(<span class="kw">getNumInd</span>(msats.fine), <span class="dv">10</span>), , ]
sub.msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 10 samples, 5 loci, 2 strata
## 
## Strata summary:
##    stratum num.ind num.missing num.alleles
## 1  Coastal       7           0         3.6
## 2 Offshore       3           0         4.4</code></pre>
<p>…or to return specific loci:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>sub.msats[, <span class="kw">c</span>(<span class="st">&quot;D11t&quot;</span>, <span class="st">&quot;EV37&quot;</span>, <span class="st">&quot;TV7&quot;</span>), ]
sub.msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 10 samples, 2 loci, 2 strata
## 
## Strata summary:
##    stratum num.ind num.missing num.alleles
## 1  Coastal       7           0         3.0
## 2 Offshore       3           0         4.5</code></pre>
<p>…or some loci in a specific stratum:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>msats.fine[, <span class="kw">c</span>(<span class="st">&quot;Ttr11&quot;</span>, <span class="st">&quot;D11t&quot;</span>), <span class="st">&quot;Coastal&quot;</span>]
sub.msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2020-02-03 22:30:55 &gt;&gt;&gt;
## 
## Contents: 68 samples, 2 loci, 1 stratum
## 
## Strata summary:
##   stratum num.ind num.missing num.alleles
## 1 Coastal      68         0.5         3.5</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>Several functions have been defined for <em>gtypes</em>, that provide summaries for individuals (<code>summarizeInds()</code>), loci (<code>summarizeLoci()</code>), and sequences (<code>summarizeSeqs()</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarizeLoci</span>(msats.fine)</code></pre></div>
<pre><code>##   locus num.genotyped num.missing prop.genotyped num.alleles allelic.richness
## 1  D11t           125           1      0.9920635          12       0.09600000
## 2  EV37           119           7      0.9444444          22       0.18487395
## 3  EV94           125           1      0.9920635          15       0.12000000
## 4 Ttr11           125           1      0.9920635           9       0.07200000
## 5 Ttr34           126           0      1.0000000          10       0.07936508
##   prop.unique.alleles exptd.het obsvd.het
## 1          0.03200000 0.7468273 0.6984127
## 2          0.02521008 0.8270751 0.6587302
## 3          0.01600000 0.8304900 0.7698413
## 4          0.02400000 0.7953414 0.6984127
## 5          0.01587302 0.8144248 0.6984127</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarizeInds</span>(msats.fine)</code></pre></div>
<pre><code>##        id  stratum num.loci.genotyped num.loci.missing.genotypes
## 1   23945  Coastal                  5                          0
## 2   25503  Coastal                  5                          0
## 3   25509  Coastal                  5                          0
## 4   40915  Coastal                  5                          0
## 5   40916  Coastal                  4                          1
## 6   41538  Coastal                  5                          0
## 7   41539  Coastal                  5                          0
## 8   41540  Coastal                  5                          0
## 9   41578  Coastal                  5                          0
## 10  41579  Coastal                  4                          1
## 11  41819  Coastal                  5                          0
## 12  41820  Coastal                  5                          0
## 13  41821  Coastal                  5                          0
## 14  41822  Coastal                  5                          0
## 15  42192  Coastal                  4                          1
## 16  42193  Coastal                  4                          1
## 17  44718  Coastal                  5                          0
## 18  44719  Coastal                  5                          0
## 19  44720  Coastal                  5                          0
## 20  44721  Coastal                  5                          0
## 21  45229  Coastal                  5                          0
## 22  45230  Coastal                  5                          0
## 23  45231  Coastal                  5                          0
## 24  45232  Coastal                  5                          0
## 25  45233  Coastal                  5                          0
## 26  45234  Coastal                  5                          0
## 27  45236  Coastal                  5                          0
## 28  45237  Coastal                  4                          1
## 29  49095  Coastal                  5                          0
## 30  51981  Coastal                  4                          1
## 31  51982  Coastal                  5                          0
## 32  78033  Coastal                  5                          0
## 33  78034  Coastal                  5                          0
## 34  78035  Coastal                  5                          0
## 35  78036  Coastal                  5                          0
## 36  78037  Coastal                  5                          0
## 37  78038  Coastal                  5                          0
## 38  78039  Coastal                  5                          0
## 39  78040  Coastal                  5                          0
## 40  78041  Coastal                  5                          0
## 41  78042  Coastal                  5                          0
## 42  78043  Coastal                  5                          0
## 43  78044  Coastal                  5                          0
## 44  78045  Coastal                  5                          0
## 45  78046  Coastal                  5                          0
## 46  78047  Coastal                  5                          0
## 47  78048  Coastal                  5                          0
## 48  78049  Coastal                  5                          0
## 49  78050  Coastal                  5                          0
## 50  78051  Coastal                  5                          0
## 51  78052  Coastal                  5                          0
## 52  78053  Coastal                  5                          0
## 53  78054  Coastal                  5                          0
## 54  78055  Coastal                  5                          0
## 55  78056  Coastal                  5                          0
## 56  78057  Coastal                  5                          0
## 57  78058  Coastal                  5                          0
## 58  78059  Coastal                  5                          0
## 59  78060  Coastal                  5                          0
## 60  78061  Coastal                  5                          0
## 61  78062  Coastal                  5                          0
## 62  78063  Coastal                  5                          0
## 63  78064  Coastal                  5                          0
## 64  78065  Coastal                  5                          0
## 65  78066  Coastal                  5                          0
## 66  78067  Coastal                  5                          0
## 67  78068  Coastal                  5                          0
## 68  78069  Coastal                  5                          0
## 69  18650 Offshore                  5                          0
## 70  18651 Offshore                  5                          0
## 71  18652 Offshore                  5                          0
## 72  18653 Offshore                  5                          0
## 73  18654 Offshore                  5                          0
## 74  18655 Offshore                  5                          0
## 75  23792 Offshore                  5                          0
## 76  23793 Offshore                  5                          0
## 77  23794 Offshore                  5                          0
## 78  23801 Offshore                  5                          0
## 79  25182 Offshore                  5                          0
## 80  25184 Offshore                  5                          0
## 81  25185 Offshore                  5                          0
## 82  25186 Offshore                  5                          0
## 83  25469 Offshore                  5                          0
## 84  25470 Offshore                  5                          0
## 85  25471 Offshore                  5                          0
## 86  26304 Offshore                  5                          0
## 87  26305 Offshore                  5                          0
## 88  26310 Offshore                  5                          0
## 89  26316 Offshore                  5                          0
## 90  26317 Offshore                  5                          0
## 91  26318 Offshore                  5                          0
## 92  26320 Offshore                  5                          0
## 93  31888 Offshore                  5                          0
## 94  41757 Offshore                  5                          0
## 95  41758 Offshore                  5                          0
## 96  41759 Offshore                  5                          0
## 97   4495 Offshore                  3                          2
## 98   4496 Offshore                  3                          2
## 99   4498 Offshore                  5                          0
## 100 50742 Offshore                  5                          0
## 101 50743 Offshore                  5                          0
## 102 50744 Offshore                  5                          0
## 103 50745 Offshore                  5                          0
## 104 50746 Offshore                  5                          0
## 105 51382 Offshore                  5                          0
## 106 51383 Offshore                  5                          0
## 107 51384 Offshore                  5                          0
## 108  5814 Offshore                  5                          0
## 109  5815 Offshore                  5                          0
## 110  5816 Offshore                  5                          0
## 111  5817 Offshore                  5                          0
## 112  5818 Offshore                  5                          0
## 113  6151 Offshore                  5                          0
## 114  6153 Offshore                  5                          0
## 115  6290 Offshore                  5                          0
## 116 74959 Offshore                  5                          0
## 117 74960 Offshore                  5                          0
## 118 74961 Offshore                  5                          0
## 119 74962 Offshore                  5                          0
## 120 74963 Offshore                  5                          0
## 121 74964 Offshore                  5                          0
## 122 74965 Offshore                  5                          0
## 123 74966 Offshore                  5                          0
## 124 78530 Offshore                  5                          0
## 125 78531 Offshore                  5                          0
## 126 78532 Offshore                  5                          0
##     pct.loci.missing.genotypes pct.loci.homozygous
## 1                          0.0                 0.4
## 2                          0.0                 0.6
## 3                          0.0                 0.8
## 4                          0.0                 0.6
## 5                          0.2                 0.0
## 6                          0.0                 0.4
## 7                          0.0                 0.6
## 8                          0.0                 0.0
## 9                          0.0                 0.0
## 10                         0.2                 0.2
## 11                         0.0                 0.0
## 12                         0.0                 0.2
## 13                         0.0                 0.8
## 14                         0.0                 0.6
## 15                         0.2                 0.4
## 16                         0.2                 0.0
## 17                         0.0                 0.6
## 18                         0.0                 0.2
## 19                         0.0                 0.2
## 20                         0.0                 0.2
## 21                         0.0                 0.4
## 22                         0.0                 0.0
## 23                         0.0                 0.6
## 24                         0.0                 0.6
## 25                         0.0                 0.2
## 26                         0.0                 0.2
## 27                         0.0                 0.6
## 28                         0.2                 0.2
## 29                         0.0                 0.4
## 30                         0.2                 0.6
## 31                         0.0                 0.4
## 32                         0.0                 0.0
## 33                         0.0                 0.0
## 34                         0.0                 0.2
## 35                         0.0                 1.0
## 36                         0.0                 0.4
## 37                         0.0                 0.2
## 38                         0.0                 0.8
## 39                         0.0                 0.0
## 40                         0.0                 0.4
## 41                         0.0                 0.0
## 42                         0.0                 0.2
## 43                         0.0                 0.6
## 44                         0.0                 0.2
## 45                         0.0                 0.4
## 46                         0.0                 0.4
## 47                         0.0                 0.0
## 48                         0.0                 0.2
## 49                         0.0                 0.4
## 50                         0.0                 0.4
## 51                         0.0                 0.6
## 52                         0.0                 0.4
## 53                         0.0                 0.2
## 54                         0.0                 0.2
## 55                         0.0                 0.8
## 56                         0.0                 0.4
## 57                         0.0                 0.2
## 58                         0.0                 0.2
## 59                         0.0                 0.8
## 60                         0.0                 0.6
## 61                         0.0                 0.4
## 62                         0.0                 0.4
## 63                         0.0                 0.2
## 64                         0.0                 0.6
## 65                         0.0                 0.4
## 66                         0.0                 0.6
## 67                         0.0                 0.4
## 68                         0.0                 0.4
## 69                         0.0                 0.0
## 70                         0.0                 0.2
## 71                         0.0                 0.0
## 72                         0.0                 0.0
## 73                         0.0                 0.2
## 74                         0.0                 0.0
## 75                         0.0                 0.0
## 76                         0.0                 0.2
## 77                         0.0                 0.2
## 78                         0.0                 0.0
## 79                         0.0                 0.0
## 80                         0.0                 0.4
## 81                         0.0                 0.0
## 82                         0.0                 0.4
## 83                         0.0                 0.0
## 84                         0.0                 0.4
## 85                         0.0                 0.0
## 86                         0.0                 0.2
## 87                         0.0                 0.6
## 88                         0.0                 0.2
## 89                         0.0                 0.2
## 90                         0.0                 0.2
## 91                         0.0                 0.2
## 92                         0.0                 0.0
## 93                         0.0                 0.0
## 94                         0.0                 0.4
## 95                         0.0                 0.2
## 96                         0.0                 0.2
## 97                         0.4                 0.4
## 98                         0.4                 0.2
## 99                         0.0                 0.6
## 100                        0.0                 0.6
## 101                        0.0                 0.2
## 102                        0.0                 0.0
## 103                        0.0                 0.2
## 104                        0.0                 0.2
## 105                        0.0                 0.6
## 106                        0.0                 0.0
## 107                        0.0                 0.6
## 108                        0.0                 0.2
## 109                        0.0                 0.6
## 110                        0.0                 0.0
## 111                        0.0                 0.2
## 112                        0.0                 0.0
## 113                        0.0                 0.0
## 114                        0.0                 0.0
## 115                        0.0                 0.0
## 116                        0.0                 0.0
## 117                        0.0                 0.4
## 118                        0.0                 0.0
## 119                        0.0                 0.0
## 120                        0.0                 0.4
## 121                        0.0                 0.0
## 122                        0.0                 0.0
## 123                        0.0                 0.2
## 124                        0.0                 0.2
## 125                        0.0                 0.4
## 126                        0.0                 0.0</code></pre>
</div>
<div id="stratifying-samples" class="section level2">
<h2>Stratifying samples</h2>
<p>You can specify the stratification scheme when creating a <em>gtypes</em> object as in the examples above. Once a <em>gtypes</em> object has been created, you can also change the stratification scheme by either supplying a new vector for the @strata slot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># randomly stratify individuals to two populations</span>
msats &lt;-<span class="st"> </span>msats.g
new.strata &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&quot;Pop1&quot;</span>, <span class="st">&quot;Pop2&quot;</span>), <span class="kw">getNumInd</span>(msats), <span class="dt">rep =</span> <span class="ot">TRUE</span>)
<span class="kw">names</span>(new.strata) &lt;-<span class="st"> </span><span class="kw">getIndNames</span>(msats)
<span class="kw">setStrata</span>(msats) &lt;-<span class="st"> </span>new.strata
msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 2 strata
## Stratification schemes: broad, fine
## 
## Strata summary:
##   stratum num.ind num.missing num.alleles
## 1    Pop1      63         0.8        11.8
## 2    Pop2      63         1.2        12.4</code></pre>
<p>or, if there is a stratification scheme data.frame in the @schemes slot, you can use the <em>stratify</em> function to choose a stratification scheme:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># choose &quot;broad&quot; stratification scheme</span>
msats &lt;-<span class="st"> </span><span class="kw">stratify</span>(msats, <span class="st">&quot;broad&quot;</span>)
msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 2 strata
## Stratification schemes: broad, fine
## 
## Strata summary:
##    stratum num.ind num.missing num.alleles
## 1  Coastal      68         1.2         4.8
## 2 Offshore      58         0.8        13.6</code></pre>
<p>You can update the @schemes slot with data.frame like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new.schemes &lt;-<span class="st"> </span><span class="kw">getSchemes</span>(msats)
new.schemes<span class="op">$</span>ran.pop &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&quot;Pop5&quot;</span>, <span class="st">&quot;Pop6&quot;</span>), <span class="kw">getNumInd</span>(msats), <span class="dt">rep =</span> <span class="ot">TRUE</span>)
<span class="kw">setSchemes</span>(msats) &lt;-<span class="st"> </span>new.schemes</code></pre></div>
<p>NOTE: Filling or changing the @schemes slot does not affect the current stratification of the samples. You must then select a new stratification scheme or fill the @strata slot as above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stratify</span>(msats, <span class="st">&quot;ran.pop&quot;</span>)</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 2 strata
## Stratification schemes: broad, fine, ran.pop
## 
## Strata summary:
##   stratum num.ind num.missing num.alleles
## 1    Pop5      59         1.6        11.2
## 2    Pop6      67         0.4        13.0</code></pre>
<p>If some samples should be unstratified (excluded from any stratified analyses), they should have NAs in the appropriate position in the @strata slot. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># unstratify a random 10 samples</span>
x &lt;-<span class="st"> </span><span class="kw">getStrata</span>(msats)
x[<span class="kw">sample</span>(<span class="kw">getIndNames</span>(msats), <span class="dv">10</span>)] &lt;-<span class="st"> </span><span class="ot">NA</span>
msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 2 strata
## Stratification schemes: broad, fine, ran.pop
## 
## Strata summary:
##    stratum num.ind num.missing num.alleles
## 1  Coastal      68         1.2         4.8
## 2 Offshore      58         0.8        13.6</code></pre>
<p>You can also randomly permute the current stratification scheme using the <em>permuteStrata()</em> function like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">msats &lt;-<span class="st"> </span><span class="kw">stratify</span>(msats, <span class="st">&quot;fine&quot;</span>)

<span class="co"># original</span>
msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 3 strata
## Stratification schemes: broad, fine, ran.pop
## 
## Strata summary:
##          stratum num.ind num.missing num.alleles
## 1        Coastal      68         1.2         4.8
## 2 Offshore.North      40         0.8        12.6
## 3 Offshore.South      18         0.0        11.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># permuted</span>
ran.msats &lt;-<span class="st"> </span><span class="kw">permuteStrata</span>(msats)
ran.msats</code></pre></div>
<pre><code>## 
## &lt;&lt;&lt; dolphin msats &gt;&gt;&gt;
## 
## Contents: 126 samples, 5 loci, 3 strata
## Stratification schemes: broad, fine, ran.pop
## 
## Strata summary:
##          stratum num.ind num.missing num.alleles
## 1        Coastal      68         0.6        12.6
## 2 Offshore.North      40         1.2        10.2
## 3 Offshore.South      18         0.2         9.0</code></pre>
<p>NOTE: Only samples assigned to strata are permuted with <em>permuteStrata</em>. Those not assigned (NAs) remain unassigned.</p>
</div>
<div id="exporting" class="section level2">
<h2>Exporting</h2>
<p>The allelic data in a <em>gtypes</em> object can be converted back to a matrix or data frame with <em>as.matrix()</em> and <em>as.data.frame()</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gen.mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(msats)
<span class="kw">head</span>(gen.mat)</code></pre></div>
<pre><code>##      id      stratum          D11t.1 D11t.2 EV37.1 EV37.2 EV94.1 EV94.2 Ttr11.1
## [1,] &quot;18650&quot; &quot;Offshore.South&quot; &quot;131&quot;  &quot;133&quot;  &quot;212&quot;  &quot;222&quot;  &quot;263&quot;  &quot;265&quot;  &quot;197&quot;  
## [2,] &quot;18651&quot; &quot;Offshore.South&quot; &quot;137&quot;  &quot;143&quot;  &quot;200&quot;  &quot;228&quot;  &quot;249&quot;  &quot;251&quot;  &quot;197&quot;  
## [3,] &quot;18652&quot; &quot;Offshore.South&quot; &quot;131&quot;  &quot;133&quot;  &quot;212&quot;  &quot;218&quot;  &quot;249&quot;  &quot;251&quot;  &quot;211&quot;  
## [4,] &quot;18653&quot; &quot;Offshore.South&quot; &quot;133&quot;  &quot;139&quot;  &quot;202&quot;  &quot;220&quot;  &quot;229&quot;  &quot;245&quot;  &quot;197&quot;  
## [5,] &quot;18654&quot; &quot;Offshore.South&quot; &quot;131&quot;  &quot;135&quot;  &quot;214&quot;  &quot;216&quot;  &quot;249&quot;  &quot;249&quot;  &quot;197&quot;  
## [6,] &quot;18655&quot; &quot;Offshore.South&quot; &quot;131&quot;  &quot;137&quot;  &quot;222&quot;  &quot;254&quot;  &quot;243&quot;  &quot;255&quot;  &quot;207&quot;  
##      Ttr11.2 Ttr34.1 Ttr34.2
## [1,] &quot;209&quot;   &quot;185&quot;   &quot;187&quot;  
## [2,] &quot;197&quot;   &quot;185&quot;   &quot;187&quot;  
## [3,] &quot;213&quot;   &quot;185&quot;   &quot;191&quot;  
## [4,] &quot;215&quot;   &quot;189&quot;   &quot;195&quot;  
## [5,] &quot;211&quot;   &quot;185&quot;   &quot;187&quot;  
## [6,] &quot;211&quot;   &quot;183&quot;   &quot;193&quot;</code></pre>
<p>By default, this function splits each allele into its own column. One can make a matrix with one locus per column and alleles separated by a specified character by setting the <em>one.col</em> argument to <em>TRUE</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gen.mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(msats, <span class="dt">one.col =</span> <span class="ot">TRUE</span>)
<span class="kw">head</span>(gen.mat)</code></pre></div>
<pre><code>##      id      stratum          D11t      EV37      EV94      Ttr11     Ttr34    
## [1,] &quot;18650&quot; &quot;Offshore.South&quot; &quot;131/133&quot; &quot;212/222&quot; &quot;263/265&quot; &quot;197/209&quot; &quot;185/187&quot;
## [2,] &quot;18651&quot; &quot;Offshore.South&quot; &quot;137/143&quot; &quot;200/228&quot; &quot;249/251&quot; &quot;197/197&quot; &quot;185/187&quot;
## [3,] &quot;18652&quot; &quot;Offshore.South&quot; &quot;131/133&quot; &quot;212/218&quot; &quot;249/251&quot; &quot;211/213&quot; &quot;185/191&quot;
## [4,] &quot;18653&quot; &quot;Offshore.South&quot; &quot;133/139&quot; &quot;202/220&quot; &quot;229/245&quot; &quot;197/215&quot; &quot;189/195&quot;
## [5,] &quot;18654&quot; &quot;Offshore.South&quot; &quot;131/135&quot; &quot;214/216&quot; &quot;249/249&quot; &quot;197/211&quot; &quot;185/187&quot;
## [6,] &quot;18655&quot; &quot;Offshore.South&quot; &quot;131/137&quot; &quot;222/254&quot; &quot;243/255&quot; &quot;207/211&quot; &quot;183/193&quot;</code></pre>
<p>The contents of a <em>gtypes</em> object can be written to a file with the <em>writeGtypes()</em> function. This will write a .csv file with the allelic information and a .fasta file for any sequence data if it exists.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
