<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Eric Archer" />

<meta name="date" content="2016-08-21" />

<title>QA/QC</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">QA/QC</h1>
<h4 class="author"><em>Eric Archer</em></h4>
<h4 class="date"><em>2016-08-21</em></h4>



<p>Quality of genotypes and sequences can be assessed with several functions. For genotypic data, by-locus summaries can be conducted with the <em>summarizeLoci</em> function, which will also produce summaries for each strata:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(msats.g)
msats &lt;-<span class="st"> </span>msats.g
smry &lt;-<span class="st"> </span><span class="kw">summarizeLoci</span>(msats)
<span class="kw">head</span>(smry)</code></pre></div>
<pre><code>##       num.genotyped prop.genotyped num.alleles allelic.richness
## D11t            125           0.99          12            0.096
## EV37            119           0.94          22            0.185
## EV94            125           0.99          15            0.120
## Ttr11           125           0.99           9            0.072
## Ttr34           126           1.00          10            0.079
##       prop.unique.alleles exptd.heterozygosity obsvd.heterozygosity
## D11t                0.250                 0.75                 0.70
## EV37                0.136                 0.83                 0.70
## EV94                0.067                 0.83                 0.78
## Ttr11               0.222                 0.80                 0.70
## Ttr34               0.200                 0.81                 0.70</code></pre>
<p>The <em>dupGenotypes</em> function identifies samples that have the same or nearly the same genotypes. The number (or percent) of loci that must be shared in order for it to be considered a duplicate can be set by the <em>num.shared</em> argument. The return data.frame provides which loci the two samples show mismatches at so they can be reviewed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Find samples that share alleles at 2/3rds of the loci</span>
<span class="kw">dupGenotypes</span>(msats, <span class="dt">num.shared =</span> <span class="fl">0.66</span>)</code></pre></div>
<pre><code>##    ids.1 ids.2       strata.1 strata.2 num.loci.genotyped num.loci.shared
## 1  41579 45237        Coastal  Coastal                  4               4
## 2  23945 78065        Coastal  Coastal                  5               4
## 3  78034 78040        Coastal  Coastal                  5               4
## 4  78034 78043        Coastal  Coastal                  5               4
## 5  78035 78053        Coastal  Coastal                  5               4
## 6  78045 78058        Coastal  Coastal                  5               4
## 7  25503 78053        Coastal  Coastal                  5               4
## 8  25509 41822        Coastal  Coastal                  5               4
## 9  41540 78040        Coastal  Coastal                  5               4
## 10 41578 45233        Coastal  Coastal                  5               4
## 11 44720 78058        Coastal  Coastal                  5               4
## 12 45230 78040        Coastal  Coastal                  5               4
## 13 40916 78038        Coastal  Coastal                  4               3
## 14 41820 42193        Coastal  Coastal                  4               3
## 15 42193 44719        Coastal  Coastal                  4               3
## 16 42193 45230        Coastal  Coastal                  4               3
## 17 42193 78035        Coastal  Coastal                  4               3
## 18 42193 78045        Coastal  Coastal                  4               3
## 19  4495 41821 Offshore.North  Coastal                  3               2
## 20  4495 78041 Offshore.North  Coastal                  3               2
## 21  4495 78054 Offshore.North  Coastal                  3               2
## 22  4495 78056 Offshore.North  Coastal                  3               2
##    prop.loci.shared mismatch.loci
## 1              1.00              
## 2              0.80          EV37
## 3              0.80          EV94
## 4              0.80          EV37
## 5              0.80          D11t
## 6              0.80         Ttr34
## 7              0.80         Ttr11
## 8              0.80         Ttr34
## 9              0.80          D11t
## 10             0.80          EV94
## 11             0.80          EV94
## 12             0.80         Ttr11
## 13             0.75         Ttr34
## 14             0.75         Ttr11
## 15             0.75          EV94
## 16             0.75          EV94
## 17             0.75         Ttr34
## 18             0.75          EV94
## 19             0.67         Ttr11
## 20             0.67         Ttr11
## 21             0.67         Ttr11
## 22             0.67         Ttr11</code></pre>
<p>The start and end positions and number of N’s and indels can be generated with the <em>summarizeSeqs</em> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(dolph.seqs)
seq.smry &lt;-<span class="st"> </span><span class="kw">summarizeSeqs</span>(<span class="kw">as.DNAbin</span>(dolph.seqs))
<span class="kw">head</span>(seq.smry)</code></pre></div>
<pre><code>##      start end length num.ns num.indels
## 4495     1 402    402      0          2
## 4496     1 402    402      0          2
## 4498     1 402    402      0          1
## 5814     1 402    402      0          2
## 5815     1 402    402      0          2
## 5816     1 402    402      0          2</code></pre>
<p>Base frequencies can be generated with <em>baseFreqs</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bf &lt;-<span class="st"> </span><span class="kw">baseFreqs</span>(<span class="kw">as.DNAbin</span>(dolph.seqs))

<span class="co"># nucleotide frequencies by site</span>
bf$site.freq[, <span class="dv">1</span>:<span class="dv">15</span>]</code></pre></div>
<pre><code>##     1   2   3   4   5   6 7   8   9  10  11  12  13  14  15
## a   0 126 126 126 126 126 5   0   0   0   0 126   0   0   0
## c   0   0   0   0   0   0 0   0 126   0   0   0   0   0   0
## g 126   0   0   0   0   0 0 126   0   0   0   0   0   0 126
## t   0   0   0   0   0   0 0   0   0 126 126   0 126 126   0
## u   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## r   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## y   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## m   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## k   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## w   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## s   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## b   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## d   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## h   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## v   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## n   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## x   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## -   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0
## .   0   0   0   0   0   0 0   0   0   0   0   0   0   0   0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># overall nucleotide frequencies</span>
bf$base.freqs</code></pre></div>
<pre><code>## 
##      a      c      g      t      u      r      y      m      k      w 
## 0.2997 0.2282 0.1283 0.3389 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 
##      s      b      d      h      v      n      x      -      . 
## 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0048 0.0000</code></pre>
<p>Sequences can be scanned for low-frequency substitutions with <em>lowFreqSubs</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lowFreqSubs</span>(<span class="kw">as.DNAbin</span>(dolph.seqs), <span class="dt">min.freq =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>##      id site base       motif
## 1 74962   57    a taaaaataatt
## 2 74962  104    g catacgcatgt
## 3 74962  392    t catgctccgtg
## 4 74962  393    c atgctccgtgg
## 5 23792  274    t cctattgatcc
## 6 26304  274    a cctataaatcc
## 7 26304  394    t taccttgtggg
## 8 23794  287    g cctccgttata</code></pre>
<p>Unusual sequences can be identified by plotting likelihoods based on pairwise distances:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(dolph.haps)
<span class="kw">sequenceLikelihoods</span>(<span class="kw">as.DNAbin</span>(dolph.haps))</code></pre></div>
<pre><code>##            id logLik
## Hap.01 Hap.01   -1.7
## Hap.02 Hap.02  -16.0
## Hap.03 Hap.03   -8.9
## Hap.04 Hap.04   -8.8
## Hap.05 Hap.05   -2.9
## Hap.06 Hap.06  -19.2
## Hap.07 Hap.07   -6.8
## Hap.08 Hap.08   -1.4
## Hap.09 Hap.09   -7.5
## Hap.10 Hap.10  -11.1
## Hap.11 Hap.11    0.0
## Hap.12 Hap.12   -7.4
## Hap.13 Hap.13   -4.3
## Hap.14 Hap.14   -7.6
## Hap.15 Hap.15  -14.8
## Hap.16 Hap.16   -2.9
## Hap.17 Hap.17   -2.2
## Hap.18 Hap.18   -7.2
## Hap.19 Hap.19   -7.0
## Hap.20 Hap.20   -3.3
## Hap.21 Hap.21   -4.3
## Hap.22 Hap.22  -20.7
## Hap.23 Hap.23   -9.2
## Hap.24 Hap.24   -2.7
## Hap.25 Hap.25   -1.8
## Hap.26 Hap.26   -3.3
## Hap.27 Hap.27   -3.2
## Hap.28 Hap.28   -1.2
## Hap.29 Hap.29  -13.4
## Hap.30 Hap.30   -9.4
## Hap.31 Hap.31   -8.1
## Hap.32 Hap.32  -26.3
## Hap.33 Hap.33   -8.8</code></pre>
<p>All of the above functions can be conducted at once with the <em>qaqc</em> function. Only those functions appropriate to the data type contained (haploid or diploid) will be run. Files are written for each output that are labelled either by the <em>@description</em> slot of the gtypes object or the optional <em>label</em> argument of the function.</p>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
