---
title: "strataG func test"
author: "Eric Archer"
date: "9/18/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  comment = NA,
  cache = TRUE
)
```

```{r}
rm(list = ls())
library(tidyverse)
library(strataG)

rm(list = ls())
ms <- read.csv("dolph.msats.csv", stringsAsFactors = FALSE)
st <- read.csv("dolph.strata.csv", stringsAsFactors = FALSE) 

ms.g <- new("gtypes", gen.data = ms[, -1], ploidy = 2,
            ind.names = ms[, 1], schemes = st)
ms.g

msats.df <- st %>% 
  select(id, broad) %>% 
  left_join(ms, by = "id")

ms.df.g <- df2gtypes(msats.df, 2, schemes = st)
ms.df.g
```


Test accessors
```{r}
getNumInd(ms.g)
getNumLoci(ms.g)
getNumStrata(ms.g)

getIndNames(ms.g)
getLocusNames(ms.g)
getStrataNames(ms.g)

strata(ms.g)
ploidy(ms.g)
other(ms.g)
description(ms.g)
```

```{r}
ms.g <- stratify(ms.g, "fine")
getNumInd(ms.g)
getNumLoci(ms.g)
getNumStrata(ms.g)

getIndNames(ms.g)
getLocusNames(ms.g)
getAlleleNames(ms.g)
getStrataNames(ms.g)

getNumInd(ms.g, T)
getIndNames(ms.g, T)
```

```{r}
strata(ms.g)
permuteStrata(ms.g)
```

Test indexing
```{r}
ms.g[, , "Coastal"]
ms.g[, c("Ttr48", "TV7"), ]
ms.g[c("23945", "25470", "25182"), , ]
```

Test sequence gtypes and accessors
```{r}
cr <- read.FASTA("dolph.haps.fasta") 
cr.g <- new("gtypes", gen.data = st$dLoop, ploidy = 1,
            ind.names = st$id, schemes = st, sequences = cr)
cr.g

cr.g <- stratify(cr.g, "broad")
cr.g
sequences(cr.g)
sequences(cr.g, F)
```

Allele frequencies
```{r}
alleleFreqs(ms.g)
alleleFreqs(ms.g, type = "prop")
alleleFreqs(ms.g, T)
alleleFreqs(ms.g, T, "prop")

alleleFreqs(cr.g)
alleleFreqs(cr.g, type = "prop")
alleleFreqs(cr.g, T)
alleleFreqs(cr.g, T, "prop")
```


Allelic richness
```{r}
allelicRichness(ms.g)
allelicRichness(ms.g, T)
```


as.data.frame and as.matrix
```{r}
as.data.frame(ms.g)
as.matrix(ms.g, one.col = TRUE)
```

Heterozygosity
```{r}
heterozygosity(ms.g, T, "exp")
heterozygosity(ms.g, type = "observed")

heterozygosity(cr.g, T, type = "expected")
```

Label Haplotypes
```{r}
library(ape)
cr.seqs <- read.FASTA("dolph.seqs.fasta")

labelHaplotypes(cr.seqs)

cr.g <- sequence2gtypes(cr.seqs, schemes = st)
cr.g
labelHaplotypes(cr.g, prefix = "Hap.")$gtypes 
```

```{r}
privateAlleles(ms.g)
privateAlleles(cr.g)
cr.g <- stratify(cr.g, "fine")
privateAlleles(cr.g)
```





```{r}
save.image("strataG func test ws.rdata")
```

