<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Eric Archer" />

<meta name="date" content="2015-07-29" />

<title>Creating and Manipulating gtypes Objects</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Creating and Manipulating gtypes Objects</h1>
<h4 class="author"><em>Eric Archer</em></h4>
<h4 class="date"><em>2015-07-29</em></h4>
</div>


<div id="construction" class="section level2">
<h2>Construction</h2>
<p>For most functions in <em>strataG</em>, you will need to load your data into a <em>gtypes</em> object. A <em>gtypes</em> object is an R S4 class with the following slots:</p>
<ul>
<li><strong>@loci</strong> a data.frame containing the allelic data as one column per locus. Alleles are on multiple rows per column with samples listed in the same order for each allele. rownames are sample names plus allele number formatted as 12345.1 and 12345.2 where 12345 is the sample name and 1 and 2 are the first and second alleles. colnames are unique locus names.</li>
<li><strong>@sequences</strong> a <em>multidna</em> object, which is a list of <em>DNAbin</em> objects (from the <em>apex</em> and <em>ape</em> packages, respectively).</li>
<li><strong>@ploidy</strong> integer representing the ploidy of the data. There are ploidy * the number of samples rows in ‘loci’.</li>
<li><strong>@strata</strong> a factor or vector that can be coerced as to a factor as long as the number of samples representing the current stratification scheme.</li>
<li><strong>@schemes</strong> schemes a data.frame with stratification schemes in each column. Sample names are in the rownames and must match the first part of the sample names (rownames) of the ‘loci’ slot. Each column is a factor.</li>
<li><strong>@description</strong> a label for the object.</li>
<li><strong>@other</strong> a slot to carry other optional related data - unused in package analyses.</li>
</ul>
<div id="new-constructor" class="section level3">
<h3><em>new</em> constructor</h3>
<p>A gtypes object can be created with the standard S4 <em>new</em> constructor function, like this:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(strataGdevel)
<span class="co">#--- create a diploid (microsatellite) gtypes object</span>
<span class="kw">data</span>(dolph.msats)
<span class="kw">data</span>(dolph.strata)
strata.schemes &lt;-<span class="st"> </span>dolph.strata[, <span class="kw">c</span>(<span class="st">&quot;broad&quot;</span>, <span class="st">&quot;fine&quot;</span>)]
<span class="kw">rownames</span>(strata.schemes) &lt;-<span class="st"> </span>dolph.strata$id
msats &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;gtypes&quot;</span>, <span class="dt">gen.data =</span> dolph.msats[, -<span class="dv">1</span>], <span class="dt">ploidy =</span> <span class="dv">2</span>,
  <span class="dt">ind.names =</span> dolph.msats[, <span class="dv">1</span>], <span class="dt">schemes =</span> strata.schemes)

<span class="co">#--- create a haploid sequence (mtDNA) gtypes object</span>
<span class="kw">data</span>(dolph.seqs)
dloop.haps &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">haps =</span> dolph.strata$id)
<span class="kw">rownames</span>(dloop.haps) &lt;-<span class="st"> </span>dolph.strata$id
dloop &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;gtypes&quot;</span>, <span class="dt">gen.data =</span> dloop.haps, <span class="dt">ploidy =</span> <span class="dv">1</span>, 
  <span class="dt">sequences =</span> dolph.seqs)</code></pre>
<p>Both of these examples create a <em>gtypes</em> where every individual is in a single default stratum. To stratify samples when created, supply either a vector, factor, or stratification scheme name to the ‘strata’ argument in <em>new</em>, like so:</p>
<pre class="sourceCode r"><code class="sourceCode r">msats.fine &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;gtypes&quot;</span>, <span class="dt">gen.data =</span> dolph.msats[, -<span class="dv">1</span>], <span class="dt">ploidy =</span> <span class="dv">2</span>,
  <span class="dt">ind.names =</span> dolph.msats[, <span class="dv">1</span>], <span class="dt">schemes =</span> strata.schemes, <span class="dt">strata =</span> <span class="st">&quot;fine&quot;</span>)</code></pre>
</div>
<div id="df2gtypes---convert-a-data.frame-or-matrix" class="section level3">
<h3><em>df2gtypes</em> - Convert a data.frame or matrix</h3>
<p>The <em>df2gtypes</em> function assumes that you have a matrix or data.frame with columns for individual ids, stratification, and locus data. You then specify the columns in the data.frame where this information can be found.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create a single data.frame with the msat data and stratification</span>
msats.merge &lt;-<span class="st"> </span><span class="kw">merge</span>(dolph.strata, dolph.msats, <span class="dt">all.y =</span> <span class="ot">TRUE</span>, <span class="dt">description =</span> <span class="kw">date</span>())
<span class="kw">str</span>(msats.merge)</code></pre>
<pre><code>## 'data.frame':    126 obs. of  34 variables:
##  $ ids      : chr  &quot;18650&quot; &quot;18651&quot; &quot;18652&quot; &quot;18653&quot; ...
##  $ broad    : chr  &quot;Offshore&quot; &quot;Offshore&quot; &quot;Offshore&quot; &quot;Offshore&quot; ...
##  $ fine     : chr  &quot;Offshore.South&quot; &quot;Offshore.South&quot; &quot;Offshore.South&quot; &quot;Offshore.South&quot; ...
##  $ hap      : chr  &quot;Hap.06&quot; &quot;Hap.14&quot; &quot;Hap.23&quot; &quot;Hap.24&quot; ...
##  $ D11t.1   : chr  &quot;131&quot; &quot;137&quot; &quot;131&quot; &quot;133&quot; ...
##  $ D11t.2   : chr  &quot;133&quot; &quot;143&quot; &quot;133&quot; &quot;139&quot; ...
##  $ D17t.1   : chr  NA &quot;172&quot; &quot;140&quot; &quot;168&quot; ...
##  $ D17t.2   : chr  NA &quot;180&quot; &quot;170&quot; &quot;174&quot; ...
##  $ D8.1     : chr  &quot;126&quot; &quot;116&quot; &quot;114&quot; &quot;92&quot; ...
##  $ D8.2     : chr  &quot;126&quot; &quot;120&quot; &quot;128&quot; &quot;110&quot; ...
##  $ EV37.1   : chr  &quot;212&quot; &quot;200&quot; &quot;212&quot; &quot;202&quot; ...
##  $ EV37.2   : chr  &quot;222&quot; &quot;228&quot; &quot;218&quot; &quot;220&quot; ...
##  $ EV94.1   : chr  &quot;263&quot; &quot;249&quot; &quot;249&quot; &quot;229&quot; ...
##  $ EV94.2   : chr  &quot;265&quot; &quot;251&quot; &quot;251&quot; &quot;245&quot; ...
##  $ KWM12A.1 : chr  &quot;177&quot; &quot;175&quot; &quot;179&quot; &quot;177&quot; ...
##  $ KWM12A.2 : chr  &quot;183&quot; &quot;175&quot; &quot;181&quot; &quot;177&quot; ...
##  $ KWM2a.1  : chr  &quot;154&quot; &quot;146&quot; &quot;154&quot; &quot;156&quot; ...
##  $ KWM2a.2  : chr  &quot;156&quot; &quot;152&quot; &quot;156&quot; &quot;158&quot; ...
##  $ SW19t.1  : chr  &quot;118&quot; &quot;116&quot; &quot;114&quot; &quot;102&quot; ...
##  $ SW19t.2  : chr  &quot;120&quot; &quot;116&quot; &quot;116&quot; &quot;116&quot; ...
##  $ Ttr11.1  : chr  &quot;197&quot; &quot;197&quot; &quot;211&quot; &quot;197&quot; ...
##  $ Ttr11.2  : chr  &quot;209&quot; &quot;197&quot; &quot;213&quot; &quot;215&quot; ...
##  $ Ttr19.1  : chr  &quot;186&quot; &quot;190&quot; &quot;190&quot; &quot;194&quot; ...
##  $ Ttr19.2  : chr  &quot;198&quot; &quot;192&quot; &quot;192&quot; &quot;200&quot; ...
##  $ Ttr34.1  : chr  &quot;185&quot; &quot;185&quot; &quot;185&quot; &quot;189&quot; ...
##  $ Ttr34.2  : chr  &quot;187&quot; &quot;187&quot; &quot;191&quot; &quot;195&quot; ...
##  $ Ttr48.1  : chr  &quot;128&quot; &quot;138&quot; &quot;124&quot; &quot;132&quot; ...
##  $ Ttr48.2  : chr  &quot;140&quot; &quot;140&quot; &quot;128&quot; &quot;132&quot; ...
##  $ Ttr58.1  : chr  &quot;187&quot; &quot;185&quot; &quot;179&quot; &quot;165&quot; ...
##  $ Ttr58.2  : chr  &quot;189&quot; &quot;187&quot; &quot;187&quot; &quot;179&quot; ...
##  $ TtrRC11.1: chr  &quot;150&quot; &quot;150&quot; &quot;148&quot; &quot;138&quot; ...
##  $ TtrRC11.2: chr  &quot;162&quot; &quot;166&quot; &quot;162&quot; &quot;168&quot; ...
##  $ TV7.1    : chr  &quot;166&quot; &quot;166&quot; &quot;166&quot; &quot;166&quot; ...
##  $ TV7.2    : chr  &quot;168&quot; &quot;166&quot; &quot;172&quot; &quot;166&quot; ...</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># creat the gtypes object</span>
msats.fine &lt;-<span class="st"> </span><span class="kw">df2gtypes</span>(msats.merge, <span class="dt">ploidy =</span> <span class="dv">2</span>, <span class="dt">id.col =</span> <span class="dv">1</span>, <span class="dt">strata.col =</span> <span class="dv">3</span>, <span class="dt">loc.col =</span> <span class="dv">5</span>)</code></pre>
</div>
<div id="sequence2gtypes---convert-dna-sequences" class="section level3">
<h3><em>sequence2gtypes</em> - Convert DNA sequences</h3>
<p>The sequence2gtypes function assumes that you have some DNA sequences and perhaps a stratification scheme.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract and name the stratification scheme</span>
strata &lt;-<span class="st"> </span>dolph.strata$fine
<span class="kw">names</span>(strata) &lt;-<span class="st"> </span>dolph.strata$ids

<span class="co"># create the gtypes object</span>
dloop.fine &lt;-<span class="st"> </span><span class="kw">sequence2gtypes</span>(dolph.seqs, strata, <span class="dt">seq.names =</span> <span class="st">&quot;dLoop&quot;</span>,
  <span class="dt">description =</span> <span class="st">&quot;dLoop: fine-scale stratification&quot;</span>)</code></pre>
</div>
<div id="genind2gtypes---convert-genind-object" class="section level3">
<h3><em>genind2gtypes</em> - Convert genind object</h3>
<p>This converts from a <em>genind</em> object from the <em>adegenet</em> package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(adegenet)</code></pre>
<pre><code>## Loading required package: ade4
## 
##    /// adegenet 2.0.1 is loaded ////////////
## 
##    &gt; overview: '?adegenet'
##    &gt; tutorials/doc/questions: 'adegenetWeb()' 
##    &gt; bug reports/feature resquests: adegenetIssues()</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># from example(df2genind)</span>
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">locusA=</span><span class="kw">c</span>(<span class="st">&quot;11&quot;</span>,<span class="st">&quot;11&quot;</span>,<span class="st">&quot;12&quot;</span>,<span class="st">&quot;32&quot;</span>),
                 <span class="dt">locusB=</span><span class="kw">c</span>(<span class="ot">NA</span>,<span class="st">&quot;34&quot;</span>,<span class="st">&quot;55&quot;</span>,<span class="st">&quot;15&quot;</span>),
                 <span class="dt">locusC=</span><span class="kw">c</span>(<span class="st">&quot;22&quot;</span>,<span class="st">&quot;22&quot;</span>,<span class="st">&quot;21&quot;</span>,<span class="st">&quot;22&quot;</span>))
<span class="kw">row.names</span>(df) &lt;-<span class="st"> </span><span class="kw">.genlab</span>(<span class="st">&quot;genotype&quot;</span>,<span class="dv">4</span>)
obj &lt;-<span class="st"> </span><span class="kw">df2genind</span>(df, <span class="dt">ploidy=</span><span class="dv">2</span>, <span class="dt">ncode=</span><span class="dv">1</span>)
obj</code></pre>
<pre><code>## /// GENIND OBJECT /////////
## 
##  // 4 individuals; 3 loci; 9 alleles; size: 6.1 Kb
## 
##  // Basic content
##    @tab:  4 x 9 matrix of allele counts
##    @loc.n.all: number of alleles per locus (range: 2-4)
##    @loc.fac: locus factor for the 9 columns of @tab
##    @all.names: list of allele names for each locus
##    @ploidy: ploidy of each individual  (range: 2-2)
##    @type:  codom
##    @call: df2genind(X = df, ncode = 1, ploidy = 2)
## 
##  // Optional content
##    - empty -</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert to gtypes</span>
gi.g &lt;-<span class="st"> </span><span class="kw">genind2gtypes</span>(obj)
gi.g</code></pre>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2015-07-29 12:22:29 &gt;&gt;&gt;
## 
## Contents: 4 samples, 3 loci, 1 stratum
## 
## Strata summary:
##         num.samples num.missing num.alleles pct.unique.alleles
## Default           4   0.3333333           3          0.5277778
##         heterozygosity
## Default      0.4722222
## 
## Locus summary:
##        num.genotyped num.alleles pct.unique.alleles
## locusA             4           3          0.3333333
## locusB             3           4          0.7500000
## locusC             4           2          0.5000000
##        observed.heterozygosity
## locusA               0.5000000
## locusB               0.6666667
## locusC               0.2500000</code></pre>
</div>
</div>
<div id="accessor-functions" class="section level2">
<h2>Accessor functions</h2>
<p>There are several functions for getting basic information from a <em>gtypes</em> object:</p>
<ul>
<li><strong>nInd(g)</strong> Returns the number of individuals.</li>
<li><strong>nLoc(g)</strong> Returns the number of loci.</li>
<li><strong>nStrata(g)</strong> Returns the number of strata in the current scheme.</li>
<li><strong>indNames(g)</strong> Returns the names of the individuals.</li>
<li><strong>locNames(g)</strong> Returns the names of the loci or genes.</li>
<li><strong>strataNames(g)</strong> Returns the names of the strata in the current scheme.</li>
<li><strong>ploidy(g)</strong> Returns the ploidy of each locus.</li>
<li><strong>strata(g)</strong> Returns the current strata to which each individual belongs. Can be set with strata(g) &lt;- value as below.</li>
<li><strong>schemes(g)</strong> Returns a data.frame of potential stratification schemes. Can be set with schemes(g) &lt;- value as below.</li>
<li><strong>loci(g, ids = NULL, loci = NULL)</strong> Returns a data.frame of the allelic for specified individuals and loci.</li>
<li><strong>seqNames(g)</strong> Returns the names of seat of sequences.</li>
<li><strong>sequences(g)</strong> Returns the <em>multidna</em> object in the @sequences slot.</li>
<li><strong>description(g)</strong> Returns the label describing the object.</li>
<li><strong>other(g)</strong> Returns the optional data stored in the @other slot.</li>
</ul>
</div>
<div id="subsettingindexing" class="section level2">
<h2>Subsetting/Indexing</h2>
<p>A <em>gtypes</em> object can be subset using the standard R ‘[’ indexing operation. The first argument specifies the desired individuals, the second is the loci to return, and the third is the strata. All standard R indexing operations involving numerical, character, or logical vectors work for each argument. For example, to return 10 random individuals:</p>
<pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>msats.fine[<span class="kw">sample</span>(<span class="kw">nInd</span>(msats.fine), <span class="dv">10</span>), , ]
sub.msats</code></pre>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2015-07-29 12:22:29 &gt;&gt;&gt;
## 
## Contents: 10 samples, 15 loci, 3 strata
## 
## Strata summary:
##                num.samples num.missing num.alleles pct.unique.alleles
## Coastal                  5   0.1333333    2.733333          0.2377778
## Offshore.North           3   0.0000000    3.800000          0.6433333
## Offshore.South           2   0.0000000    3.133333          0.8111111
##                heterozygosity
## Coastal             0.4100000
## Offshore.North      0.7555556
## Offshore.South      0.8000000
## 
## Locus summary:
##         num.genotyped num.alleles pct.unique.alleles
## D11t               10           9          0.6666667
## D17t               10           8          0.7500000
## D8                 10           5          0.4000000
## EV37                9           8          0.8750000
## EV94               10           7          0.5714286
## KWM12A             10           6          0.3333333
## KWM2a              10           7          0.2857143
## SW19t              10           6          0.3333333
## Ttr11              10           7          0.4285714
## Ttr19               9           7          0.4285714
## Ttr34              10           7          0.2857143
## Ttr48              10           8          0.3750000
## Ttr58              10           4          0.2500000
## TtrRC11            10           8          0.7500000
## TV7                10           4          0.2500000
##         observed.heterozygosity
## D11t                  0.7000000
## D17t                  0.6000000
## D8                    0.3000000
## EV37                  0.4444444
## EV94                  0.7000000
## KWM12A                0.8000000
## KWM2a                 0.6000000
## SW19t                 0.7000000
## Ttr11                 0.6000000
## Ttr19                 0.5555556
## Ttr34                 0.6000000
## Ttr48                 0.7000000
## Ttr58                 0.5000000
## TtrRC11               0.4000000
## TV7                   0.7000000</code></pre>
<p>…or to return specific loci:</p>
<pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>sub.msats[, <span class="kw">c</span>(<span class="st">&quot;D11t&quot;</span>, <span class="st">&quot;EV37&quot;</span>, <span class="st">&quot;TV7&quot;</span>), ]
sub.msats</code></pre>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2015-07-29 12:22:29 &gt;&gt;&gt;
## 
## Contents: 10 samples, 3 loci, 3 strata
## 
## Strata summary:
##                num.samples num.missing num.alleles pct.unique.alleles
## Coastal                  5   0.3333333    2.333333          0.2222222
## Offshore.North           3   0.0000000    3.333333          0.4888889
## Offshore.South           2   0.0000000    3.000000          0.8333333
##                heterozygosity
## Coastal             0.5000000
## Offshore.North      0.7777778
## Offshore.South      0.6666667
## 
## Locus summary:
##      num.genotyped num.alleles pct.unique.alleles observed.heterozygosity
## D11t            10           9          0.6666667               0.7000000
## EV37             9           8          0.8750000               0.4444444
## TV7             10           4          0.2500000               0.7000000</code></pre>
<p>…or some loci in a specific stratum:</p>
<pre class="sourceCode r"><code class="sourceCode r">sub.msats &lt;-<span class="st"> </span>msats.fine[, <span class="kw">c</span>(<span class="st">&quot;EV94&quot;</span>, <span class="st">&quot;SW19t&quot;</span>), <span class="st">&quot;Coastal&quot;</span>]
sub.msats</code></pre>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2015-07-29 12:22:29 &gt;&gt;&gt;
## 
## Contents: 68 samples, 2 loci, 1 stratum
## 
## Strata summary:
##         num.samples num.missing num.alleles pct.unique.alleles
## Coastal          68         0.5           6                  0
##         heterozygosity
## Coastal      0.7333187
## 
## Locus summary:
##       num.genotyped num.alleles pct.unique.alleles observed.heterozygosity
## EV94             68           5                  0               0.7352941
## SW19t            67           7                  0               0.7313433</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>A <em>summary</em> function has been defined for <em>gtypes</em>, and the return value provides the number of individuals, loci, strata, allele frequencies, a by-strata summary, and a by-locus summary:</p>
<pre class="sourceCode r"><code class="sourceCode r">msats.smry &lt;-<span class="st"> </span><span class="kw">summary</span>(msats.fine)
<span class="kw">str</span>(msats.smry)</code></pre>
<pre><code>## List of 7
##  $ num.ind     : num 126
##  $ num.loc     : int 15
##  $ num.strata  : int 3
##  $ allele.freqs:List of 15
##   ..$ D11t   : num [1:12, 1:2, 1:3] 0 0 0 0 0 0 48 83 3 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:12] &quot;117&quot; &quot;119&quot; &quot;121&quot; &quot;127&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ D17t   : num [1:18, 1:2, 1:3] 97 0 0 0 0 15 11 11 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:18] &quot;140&quot; &quot;146&quot; &quot;158&quot; &quot;162&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ D8     : num [1:13, 1:2, 1:3] 1 0 0 0 21 5 108 1 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:13] &quot;104&quot; &quot;106&quot; &quot;108&quot; &quot;110&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ EV37   : num [1:22, 1:2, 1:3] 0 0 0 0 0 0 0 1 71 33 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:22] &quot;190&quot; &quot;200&quot; &quot;202&quot; &quot;204&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ EV94   : num [1:15, 1:2, 1:3] 0 0 0 12 0 47 30 0 0 25 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:15] &quot;229&quot; &quot;239&quot; &quot;243&quot; &quot;245&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ KWM12A : num [1:9, 1:2, 1:3] 0 28 62 46 0 0 0 0 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:9] &quot;167&quot; &quot;173&quot; &quot;175&quot; &quot;177&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ KWM2a  : num [1:10, 1:2, 1:3] 0 73 16 0 0 0 0 40 0 7 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:10] &quot;146&quot; &quot;148&quot; &quot;150&quot; &quot;152&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ SW19t  : num [1:16, 1:2, 1:3] 6 0 0 0 0 15 0 5 25 19 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:16] &quot;102&quot; &quot;104&quot; &quot;106&quot; &quot;108&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ Ttr11  : num [1:9, 1:2, 1:3] 0 0 42 0 0 59 33 2 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:9] &quot;193&quot; &quot;197&quot; &quot;207&quot; &quot;209&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ Ttr19  : num [1:11, 1:2, 1:3] 0 0 0 0 0 0 12 47 72 1 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:11] &quot;172&quot; &quot;174&quot; &quot;182&quot; &quot;184&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ Ttr34  : num [1:10, 1:2, 1:3] 0 0 33 19 0 65 8 11 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:10] &quot;179&quot; &quot;183&quot; &quot;185&quot; &quot;187&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ Ttr48  : num [1:11, 1:2, 1:3] 3 0 0 0 0 38 55 40 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:11] &quot;122&quot; &quot;124&quot; &quot;128&quot; &quot;130&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ Ttr58  : num [1:9, 1:2, 1:3] 0 15 0 121 0 0 0 0 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:9] &quot;165&quot; &quot;179&quot; &quot;181&quot; &quot;185&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ TtrRC11: num [1:16, 1:2, 1:3] 0 0 0 0 4 113 0 0 0 0 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:16] &quot;138&quot; &quot;140&quot; &quot;144&quot; &quot;146&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   ..$ TV7    : num [1:7, 1:2, 1:3] 0 12 41 0 81 ...
##   .. ..- attr(*, &quot;dimnames&quot;)=List of 3
##   .. .. ..$ : chr [1:7] &quot;162&quot; &quot;164&quot; &quot;166&quot; &quot;168&quot; ...
##   .. .. ..$ : chr [1:2] &quot;freq&quot; &quot;prop&quot;
##   .. .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##  $ strata.smry : num [1:3, 1:5] 68 40 18 0.733 1.067 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;Coastal&quot; &quot;Offshore.North&quot; &quot;Offshore.South&quot;
##   .. ..$ : chr [1:5] &quot;num.samples&quot; &quot;num.missing&quot; &quot;num.alleles&quot; &quot;pct.unique.alleles&quot; ...
##  $ unstratified: int 0
##  $ locus.smry  : num [1:15, 1:8] 125 124 124 119 125 124 124 124 125 123 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:15] &quot;D11t&quot; &quot;D17t&quot; &quot;D8&quot; &quot;EV37&quot; ...
##   .. ..$ : chr [1:8] &quot;num.genotyped&quot; &quot;pct.genotyped&quot; &quot;num.alleles&quot; &quot;allelic.richness&quot; ...
##  - attr(*, &quot;description&quot;)= chr &quot;gtypes created on 2015-07-29 12:22:29&quot;
##  - attr(*, &quot;class&quot;)= chr [1:2] &quot;gtypeSummary&quot; &quot;list&quot;</code></pre>
<p>For just the by-locus summary statistics:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarizeLoci</span>(msats.fine)</code></pre>
<pre><code>##         num.genotyped pct.genotyped num.alleles allelic.richness
## D11t              125     0.9920635          12       0.09600000
## D17t              124     0.9841270          18       0.14516129
## D8                124     0.9841270          13       0.10483871
## EV37              119     0.9444444          22       0.18487395
## EV94              125     0.9920635          15       0.12000000
## KWM12A            124     0.9841270           9       0.07258065
## KWM2a             124     0.9841270          10       0.08064516
## SW19t             124     0.9841270          16       0.12903226
## Ttr11             125     0.9920635           9       0.07200000
## Ttr19             123     0.9761905          11       0.08943089
## Ttr34             126     1.0000000          10       0.07936508
## Ttr48             126     1.0000000          11       0.08730159
## Ttr58             124     0.9841270           9       0.07258065
## TtrRC11           126     1.0000000          16       0.12698413
## TV7               123     0.9761905           7       0.05691057
##         pct.unique.alleles expected.heterozygosity observed.heterozygosity
## D11t            0.25000000               0.7468273               0.7040000
## D17t            0.22222222               0.7247617               0.5645161
## D8              0.23076923               0.6703343               0.4677419
## EV37            0.13636364               0.8270751               0.6974790
## EV94            0.06666667               0.8304900               0.7760000
## KWM12A          0.11111111               0.7610030               0.7177419
## KWM2a           0.20000000               0.8092269               0.7096774
## SW19t           0.25000000               0.8438357               0.7822581
## Ttr11           0.22222222               0.7953414               0.7040000
## Ttr19           0.18181818               0.7627012               0.7317073
## Ttr34           0.20000000               0.8144248               0.6984127
## Ttr48           0.27272727               0.8257446               0.7698413
## Ttr58           0.00000000               0.6111075               0.5161290
## TtrRC11         0.12500000               0.6881996               0.5714286
## TV7             0.14285714               0.6672640               0.5853659
##             theta
## D11t    0.5563279
## D17t    0.5398255
## D8      0.4991661
## EV37    0.6164407
## EV94    0.6190022
## KWM12A  0.5669354
## KWM2a   0.6030569
## SW19t   0.6290162
## Ttr11   0.5926503
## Ttr19   0.5682064
## Ttr34   0.6069539
## Ttr48   0.6154427
## Ttr58   0.4550089
## TtrRC11 0.5125053
## TV7     0.4968742</code></pre>
</div>
<div id="stratifying-samples" class="section level2">
<h2>Stratifying samples</h2>
<p>You can specify the stratification scheme when creating a <em>gtpyes</em> object as in the examples above. Once a <em>gtypes</em> object has been created, you can also change the stratification scheme by either supplying a new vector for the @strata slot:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># randomly stratify individuals to two populations</span>
<span class="kw">strata</span>(msats) &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&quot;Pop1&quot;</span>, <span class="st">&quot;Pop2&quot;</span>), <span class="kw">nInd</span>(msats), <span class="dt">rep =</span> <span class="ot">TRUE</span>)
<span class="kw">summary</span>(msats)$strata.smry</code></pre>
<pre><code>##      num.samples num.missing num.alleles pct.unique.alleles heterozygosity
## Pop1          77   0.5333333    11.13333          0.1738358      0.6756752
## Pop2          49   1.3333333    10.93333          0.2663981      0.6516251</code></pre>
<p>or, if there is a stratification scheme data.frame in the @schemes slot, you can use the <em>stratify</em> function to choose a stratification scheme:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># choose &quot;broad&quot; stratification scheme</span>
msats &lt;-<span class="st"> </span><span class="kw">stratify</span>(msats, <span class="st">&quot;broad&quot;</span>)
<span class="kw">summary</span>(msats)$strata.smry</code></pre>
<pre><code>##          num.samples num.missing num.alleles pct.unique.alleles
## Coastal           68   0.7333333     4.20000         0.07190476
## Offshore          58   1.1333333    12.26667         0.18240260
##          heterozygosity
## Coastal       0.5505505
## Offshore      0.8036931</code></pre>
<p>If the @schemes slot is empty (NULL), you can fill it with a proper data.frame like this:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">schemes</span>(dloop) &lt;-<span class="st"> </span>strata.schemes</code></pre>
<p>NOTE: Filling or changing the @schemes slot does not affect the current stratification of the samples. You must then select a new stratification scheme or fill the @strata slot as above.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stratify</span>(dloop, <span class="st">&quot;fine&quot;</span>)</code></pre>
<pre><code>## 
## &lt;&lt;&lt; gtypes created on 2015-07-29 12:22:29 &gt;&gt;&gt;
## 
## Contents: 126 samples, 1 locus, 3 strata
## 
## Strata summary:
##                num.samples num.missing num.alleles pct.unique.alleles
## Coastal                 68           0          68                  1
## Offshore.North          40           0          40                  1
## Offshore.South          18           0          18                  1
##                heterozygosity
## Coastal                     1
## Offshore.North              1
## Offshore.South              1
## 
## Sequence summary:
##      num.seqs min.length mean.length max.length         a         c
## haps      126        402         402        402 0.3011288 0.2293531
##              g         t
## haps 0.1289702 0.3405479</code></pre>
<p>If some samples should be unstratified (excluded from any stratified analyses), they should have NAs in the appropriate position in the @strata slot. For example:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># unstratify a random 10 samples</span>
x &lt;-<span class="st"> </span><span class="kw">strata</span>(msats)
x[<span class="kw">sample</span>(<span class="kw">indNames</span>(msats), <span class="dv">10</span>)] &lt;-<span class="st"> </span><span class="ot">NA</span>
<span class="kw">strata</span>(msats) &lt;-<span class="st"> </span>x
<span class="kw">summary</span>(msats)$strata.smry</code></pre>
<pre><code>##          num.samples num.missing num.alleles pct.unique.alleles
## Coastal           61   0.7333333     4.20000         0.07190476
## Offshore          55   1.1333333    12.26667         0.19323593
##          heterozygosity
## Coastal       0.5525450
## Offshore      0.8088944</code></pre>
<p>You can also randomly permute the current stratification scheme using the <em>permuteStrata</em> function like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">msats &lt;-<span class="st"> </span><span class="kw">stratify</span>(msats, <span class="st">&quot;fine&quot;</span>)

<span class="co"># original</span>
<span class="kw">summary</span>(msats)$strata.smry</code></pre>
<pre><code>##                num.samples num.missing num.alleles pct.unique.alleles
## Coastal                 68  0.73333333    4.200000         0.07190476
## Offshore.North          40  1.06666667   11.333333         0.20056277
## Offshore.South          18  0.06666667    9.733333         0.25018519
##                heterozygosity
## Coastal             0.5505505
## Offshore.North      0.7989789
## Offshore.South      0.8139434</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># permuted</span>
ran.msats &lt;-<span class="st"> </span><span class="kw">permuteStrata</span>(msats)
<span class="kw">summary</span>(ran.msats)$strata.smry</code></pre>
<pre><code>##                num.samples num.missing num.alleles pct.unique.alleles
## Coastal                 68  0.93333333    10.80000          0.1972487
## Offshore.North          40  0.86666667    10.46667          0.2808671
## Offshore.South          18  0.06666667     8.20000          0.2827898
##                heterozygosity
## Coastal             0.6632011
## Offshore.North      0.6539564
## Offshore.South      0.7065359</code></pre>
<p>NOTE: Only samples assigned to strata are permuted with <em>permuteStrata</em>. Those not assigned (NAs) remain unassigned.</p>
</div>
<div id="exporting" class="section level2">
<h2>Exporting</h2>
<p>The allelic data in a <em>gtypes</em> object can be converted back to a matrix with <em>as.matrix</em>:</p>
<pre class="sourceCode r"><code class="sourceCode r">gen.mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(msats)
<span class="kw">head</span>(gen.mat)</code></pre>
<pre><code>##      D11t.1 D11t.2 D17t.1 D17t.2 D8.1  D8.2  EV37.1 EV37.2 EV94.1 EV94.2
## 4495 &quot;133&quot;  &quot;133&quot;  &quot;170&quot;  &quot;170&quot;  NA    NA    NA     NA     NA     NA    
## 4496 &quot;137&quot;  &quot;141&quot;  &quot;140&quot;  &quot;140&quot;  NA    NA    NA     NA     &quot;243&quot;  &quot;251&quot; 
## 4498 &quot;133&quot;  &quot;133&quot;  &quot;166&quot;  &quot;170&quot;  &quot;110&quot; &quot;114&quot; &quot;220&quot;  &quot;220&quot;  &quot;245&quot;  &quot;265&quot; 
## 5814 &quot;135&quot;  &quot;139&quot;  &quot;170&quot;  &quot;170&quot;  &quot;110&quot; &quot;126&quot; &quot;210&quot;  &quot;226&quot;  &quot;251&quot;  &quot;265&quot; 
## 5815 &quot;139&quot;  &quot;139&quot;  &quot;172&quot;  &quot;194&quot;  &quot;114&quot; &quot;114&quot; &quot;212&quot;  &quot;212&quot;  &quot;249&quot;  &quot;261&quot; 
## 5816 &quot;135&quot;  &quot;137&quot;  &quot;140&quot;  &quot;140&quot;  &quot;118&quot; &quot;118&quot; &quot;214&quot;  &quot;220&quot;  &quot;249&quot;  &quot;253&quot; 
##      KWM12A.1 KWM12A.2 KWM2a.1 KWM2a.2 SW19t.1 SW19t.2 Ttr11.1 Ttr11.2
## 4495 NA       NA       NA      NA      NA      NA      &quot;211&quot;   &quot;211&quot;  
## 4496 NA       NA       NA      NA      &quot;118&quot;   &quot;118&quot;   NA      NA     
## 4498 &quot;177&quot;    &quot;179&quot;    &quot;152&quot;   &quot;156&quot;   &quot;102&quot;   &quot;124&quot;   &quot;209&quot;   &quot;211&quot;  
## 5814 &quot;167&quot;    &quot;177&quot;    &quot;150&quot;   &quot;156&quot;   &quot;116&quot;   &quot;124&quot;   &quot;209&quot;   &quot;209&quot;  
## 5815 &quot;177&quot;    &quot;177&quot;    &quot;152&quot;   &quot;156&quot;   &quot;106&quot;   &quot;114&quot;   &quot;207&quot;   &quot;213&quot;  
## 5816 &quot;177&quot;    &quot;179&quot;    &quot;152&quot;   &quot;154&quot;   &quot;120&quot;   &quot;124&quot;   &quot;207&quot;   &quot;209&quot;  
##      Ttr19.1 Ttr19.2 Ttr34.1 Ttr34.2 Ttr48.1 Ttr48.2 Ttr58.1 Ttr58.2
## 4495 NA      NA      &quot;185&quot;   &quot;191&quot;   &quot;128&quot;   &quot;132&quot;   NA      NA     
## 4496 &quot;190&quot;   &quot;196&quot;   &quot;189&quot;   &quot;189&quot;   &quot;130&quot;   &quot;138&quot;   NA      NA     
## 4498 &quot;194&quot;   &quot;194&quot;   &quot;189&quot;   &quot;189&quot;   &quot;132&quot;   &quot;134&quot;   &quot;179&quot;   &quot;187&quot;  
## 5814 &quot;196&quot;   &quot;198&quot;   &quot;185&quot;   &quot;187&quot;   &quot;130&quot;   &quot;134&quot;   &quot;187&quot;   &quot;189&quot;  
## 5815 &quot;190&quot;   &quot;192&quot;   &quot;185&quot;   &quot;185&quot;   &quot;130&quot;   &quot;136&quot;   &quot;185&quot;   &quot;187&quot;  
## 5816 &quot;192&quot;   &quot;196&quot;   &quot;187&quot;   &quot;195&quot;   &quot;130&quot;   &quot;142&quot;   &quot;179&quot;   &quot;185&quot;  
##      TtrRC11.1 TtrRC11.2 TV7.1 TV7.2
## 4495 &quot;138&quot;     &quot;138&quot;     NA    NA   
## 4496 &quot;148&quot;     &quot;148&quot;     NA    NA   
## 4498 &quot;154&quot;     &quot;168&quot;     &quot;166&quot; &quot;166&quot;
## 5814 &quot;150&quot;     &quot;154&quot;     &quot;164&quot; &quot;166&quot;
## 5815 &quot;144&quot;     &quot;148&quot;     &quot;166&quot; &quot;172&quot;
## 5816 &quot;140&quot;     &quot;150&quot;     &quot;164&quot; &quot;166&quot;</code></pre>
<p>By default, this function splits each allele into its own column. One can make a matrix with one locus per column and alleles separated by a specified character by setting the <em>one.col</em> argument to <em>TRUE</em>:</p>
<pre class="sourceCode r"><code class="sourceCode r">gen.mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(msats, <span class="dt">one.col =</span> <span class="ot">TRUE</span>)
<span class="kw">head</span>(gen.mat)</code></pre>
<pre><code>##      D11t      D17t      D8        EV37      EV94      KWM12A    KWM2a    
## 4495 &quot;133/133&quot; &quot;170/170&quot; NA        NA        NA        NA        NA       
## 4496 &quot;137/141&quot; &quot;140/140&quot; NA        NA        &quot;243/251&quot; NA        NA       
## 4498 &quot;133/133&quot; &quot;166/170&quot; &quot;110/114&quot; &quot;220/220&quot; &quot;245/265&quot; &quot;177/179&quot; &quot;152/156&quot;
## 5814 &quot;135/139&quot; &quot;170/170&quot; &quot;110/126&quot; &quot;210/226&quot; &quot;251/265&quot; &quot;167/177&quot; &quot;150/156&quot;
## 5815 &quot;139/139&quot; &quot;172/194&quot; &quot;114/114&quot; &quot;212/212&quot; &quot;249/261&quot; &quot;177/177&quot; &quot;152/156&quot;
## 5816 &quot;135/137&quot; &quot;140/140&quot; &quot;118/118&quot; &quot;214/220&quot; &quot;249/253&quot; &quot;177/179&quot; &quot;152/154&quot;
##      SW19t     Ttr11     Ttr19     Ttr34     Ttr48     Ttr58     TtrRC11  
## 4495 NA        &quot;211/211&quot; NA        &quot;185/191&quot; &quot;128/132&quot; NA        &quot;138/138&quot;
## 4496 &quot;118/118&quot; NA        &quot;190/196&quot; &quot;189/189&quot; &quot;130/138&quot; NA        &quot;148/148&quot;
## 4498 &quot;102/124&quot; &quot;209/211&quot; &quot;194/194&quot; &quot;189/189&quot; &quot;132/134&quot; &quot;179/187&quot; &quot;154/168&quot;
## 5814 &quot;116/124&quot; &quot;209/209&quot; &quot;196/198&quot; &quot;185/187&quot; &quot;130/134&quot; &quot;187/189&quot; &quot;150/154&quot;
## 5815 &quot;106/114&quot; &quot;207/213&quot; &quot;190/192&quot; &quot;185/185&quot; &quot;130/136&quot; &quot;185/187&quot; &quot;144/148&quot;
## 5816 &quot;120/124&quot; &quot;207/209&quot; &quot;192/196&quot; &quot;187/195&quot; &quot;130/142&quot; &quot;179/185&quot; &quot;140/150&quot;
##      TV7      
## 4495 NA       
## 4496 NA       
## 4498 &quot;166/166&quot;
## 5814 &quot;164/166&quot;
## 5815 &quot;166/172&quot;
## 5816 &quot;164/166&quot;</code></pre>
<p>The contents of a <em>gtypes</em> object can be written to a file with the <em>write.gtypes</em> function. This will write a .csv file with the allelic information and a .fasta file for any sequence data if it exists.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
